### PR TITLE
bpo-40939: Remove the old parser (Part 2)

### PR DESCRIPTION
**Pull request by lysnikolaou**
https://bugs.python.org/issue40939

### CODE REVIEW

**Review by gvanrossum — APPROVED** [review]
LG assuming the tests pass.

**Comment by nnemkin** [comment]
`PyParser_ASTFromXXX` functions are public API that don't depend on the parser internals. Why are they being removed? Seems like a pointless compatibility break.

File: Include/pythonrun.h [hunk 1]
```diff
@@ -32,60 +32,9 @@ PyAPI_FUNC(int) PyRun_InteractiveLoopFlags(
     const char *filename,       /* decoded from the filesystem encoding */
     PyCompilerFlags *flags);
 
-PyAPI_FUNC(struct _mod *) PyParser_ASTFromString(
```
Comment (pablogsal) [inline Include/pythonrun.h]: We cannot remove these declarations as they are part of the public API. We need to provide new implementations

Comment (lysnikolaou) [reply Include/pythonrun.h]: What if we make macros out of them and just call the corresponding `PyPegen_AST*` function? Would that work?

Comment (pablogsal) [reply Include/pythonrun.h]: We could do this in a separate PR if you wish

Comment (pablogsal) [reply Include/pythonrun.h]: > What if we make macros out of them and just call the corresponding PyPegen_AST* function? Would that work?

We could rename the PyPegen_AST functions better I think

Comment (lysnikolaou) [reply Include/pythonrun.h]: > We could rename the PyPegen_AST functions better I think

Won't there be folks that use the `PyPegen_*` functions in 3.9, whose code will break as well in that case?

Comment (pablogsal) [reply Include/pythonrun.h]: I propose to remove these changes in this PR and dealing with the implementations in other

Comment (pablogsal) [reply Include/pythonrun.h]: > Won't there be folks that use the `PyPegen_*` functions in 3.9, whose code will break as well in that case?

Are those part of the Public API? If so, we should make them private I think

Comment (lysnikolaou) [reply Include/pythonrun.h]: > I propose to remove these changes in this PR and dealing with the implementations in other

Yup, done in https://github.com/python/cpython/pull/21005/commits/1a86551632669c4062513fae1b2f26912fd83ca0.

> Are those part of the Public API? If so, we should make them private I think

Oh, you're right, they're not.

File: Include/pythonrun.h [hunk 2]
```diff
@@ -64,25 +64,7 @@ PyAPI_FUNC(struct _mod *) PyParser_ASTFromFileObject(
     PyCompilerFlags *flags,
     int *errcode,
     PyArena *arena);
-#endif
 
-#ifndef PyParser_SimpleParseString
```
Comment (pablogsal) [inline Include/pythonrun.h]: These declarations need to be restored as well

**Review by pablogsal — CHANGES_REQUESTED** [review]

File: Include/pythonrun.h (hunk 2 above)
Comment (lysnikolaou) [reply Include/pythonrun.h]: Done.

**Comment by lysnikolaou** [comment]
I have made the requested changes; please review again

**Review by pablogsal — APPROVED** [review]
LGTM

**Comment by pablogsal** [comment]
@lysnikolaou I was going to start a PR for re-implementing/re-naming the parser functions but maybe you have already started one :)

**Comment by lysnikolaou** [comment]
> @lysnikolaou I was going to start a PR for re-implementing/re-naming the parser functions but maybe you have already started one? :)

I was just about to, but you can do it if you think reviewing my PR is going to take more time. Whatever you want!

**Comment by pablogsal** [comment]
> I was just about to, but you can do it if you think reviewing my PR is going to take more time. Whatever you want!

Go ahead if you have time now, as I will only be able to do one in a couple of hours

**Comment by gvanrossum** [comment]
Agreed that the public APIs that don't depend on types like `mod_ty` from the old parser should stay. I think they should not become macros -- they should stay real functions that just call the Pegen functions. (I don't like the "pegen" name leaking into public APIs.)

**Pull request merged**